File: 数据结构/Monotone stack.cpp
int largestRectangleArea(vector<int>& nums) {
    int n = nums.size();
    vector<int> r(n, n);
    vector<int> l(n, -1);
    stack<int> s;
    for (int i = n - 1; i >= 0; i--) {
        while (!s.empty() && nums[s.top()] >= nums[i]) s.pop();
        if (!s.empty()) r[i] = s.top();
        s.push(i);
    } // 求右边第一个小于的下标
    while (!s.empty()) s.pop();
    for (int i = 0; i <= n - 1; i++) {
        while (!s.empty() && nums[s.top()] >= nums[i]) s.pop();
        if (!s.empty()) l[i] = s.top();
        s.push(i);
    } // 求左边第一个小于的下标
    int maxx = INT_MIN;
    for (int i = 0; i <= n - 1; i++) maxx = max(maxx, nums[i] * (r[i] - l[i] - 1));
    return maxx;
}

File: Python/数据结构/Monotone stack.py
def largestRectangleArea(nums: list[int]) -> int:
    n = len(nums)
    r = [n] * n
    l = [-1] * n
    s: list[int] = []
    for i in range(n - 1, -1, -1):
        while s and nums[s[-1]] >= nums[i]:
            s.pop()
        if s:
            r[i] = s[-1]
        s.append(i)
    # 求右边第一个小于的下标

    s.clear()
    for i in range(n):
        while s and nums[s[-1]] >= nums[i]:
            s.pop()
        if s:
            l[i] = s[-1]
        s.append(i)
    # 求左边第一个小于的下标

    maxx = -(1 << 31)
    for i in range(n):
        maxx = max(maxx, nums[i] * (r[i] - l[i] - 1))
    return maxx
